### ISSUE TITLE
[BUG]: `conv2d_transpose` aborts on nvidia GPU: `ABORT: invalid cudnnStatus_t entry` (CUDA 13 / cuDNN 9.20)

### ISSUE THREAD

**Issue by alisterburt** [modular/mojo#6726]
### Bug description

`ops.conv2d_transpose` hard-aborts the process on an NVIDIA GPU with the message `ABORT: invalid cudnnStatus_t entry`. 

Regular `ops.conv2d` on the same device/session works correctly, so this is specific to the transposed-conv. 

MAX is receiving a `cudnnStatus_t` value it cannot map to its enum, suggests a cuDNN version-skew.

We should confirm the supported cuDNN/CUDA matrix for MAX 26.2. Is CUDA 13 / cuDNN 9.20 supported for `conv2d_transpose`, and if not, what is? This cuda/cudnn came in with the current torch afaik so this is relevant for interop with torch

### Steps to reproduce

```python
import numpy as np, torch
import torch.nn.functional as F
from max.graph import Graph, TensorType, DeviceRef, ops
from max.dtype import DType
from max.engine import InferenceSession
from max.driver import Accelerator, CPU, Buffer

dev, host, GDEV = Accelerator(), CPU(), DeviceRef.GPU()

# small ConvTranspose2d along H: (Cin,Cout,kH,1), stride (4,1), no pad
N, Cin, H, W, Cout, kH, s = 1, 6, 8, 5, 4, 8, 4
x = np.random.randn(N, Cin, H, W).astype(np.float32)
w = np.random.randn(Cin, Cout, kH, 1).astype(np.float32)
ref = F.conv_transpose2d(torch.from_numpy(x), torch.from_numpy(w), stride=(s, 1)).numpy()

with Graph("cTg", input_types=[TensorType(DType.float32, (N, Cin, H, W), device=GDEV)]) as g:
    (xx,) = g.inputs
    xn = ops.permute(xx.tensor, [0, 2, 3, 1])                       # NHWC
    wc = ops.constant(np.ascontiguousarray(np.transpose(w, (2, 3, 1, 0))),
                      DType.float32, device=GDEV)                   # RSCF
    yy = ops.conv2d_transpose(xn, wc, stride=(s, 1))
    g.output(ops.permute(yy, [0, 3, 1, 2]))

m = InferenceSession(devices=[dev]).load(g)
b = Buffer.from_numpy(x).to(dev)
out = [o.to(host).to_numpy() for o in m.execute(b)][0]
print(np.abs(out - ref).max())
```

### System information

```shell
- MAX / modular: **26.2.0** (`max`, `max-core`, `max-mojo-libs`, `modular` all 26.2.0)
- GPU: **NVIDIA A10** (driver 580.105.08)
- CUDA: **13.0** (`cuda-toolkit` 13.0.2, `nvidia-cuda-runtime` 13.0.96)
- cuDNN: **9.20.0.48** (`nvidia-cudnn-cu13`, pulled in by torch 2.12.1+cu130)
- cuBLAS: `nvidia-cublas` 13.1.1.3
- OS: Linux x86_64, Python 3.13
```

assisted by: claude

**Comment by BradLarson** [comment]
I do see the same thing on the DGX Spark, and have a kinda hacky patch locally that works around this. Will see if that can be generalized to fix this issue.

**Comment by alisterburt** [comment]
awesome - quick response much appreciate, especially on a Saturday!

**Comment by vedmalex** [comment]
Adding two data points to this thread: **`ops.conv2d_transpose` GPU dispatch is also broken on MAX 26.5.0** (newer than the 26.2.0 in the original report), and it fails on **Apple Silicon / Metal** as well — which I haven't seen covered in any open issue. Both are the same "transposed conv specifically, while plain `ops.conv2d` is fine on the same device/session" shape @alisterburt reported. @BradLarson — if your local workaround patch is about the cuDNN status/dispatch path, these may tell you whether it generalizes.

### 1. Metal: not a status-mapping problem but a missing symbol — CUDA-only code path on a non-CUDA backend

On an M1, `ops.conv2d_transpose` on `Accelerator()` **aborts the process**:

```
ABORT: oss/modular/mojo/stdlib/std/ffi/__init__.mojo:762:18: symbol not found: cudnnCreate
```

That's **NVIDIA cuDNN being dynamically loaded on an Apple Metal device**. So the GPU dispatch for this op appears to take a CUDA/cuDNN path unconditionally, without branching on the actual accelerator backend. This is uncatchable — `try`/`except` around `model.execute(...)` doesn't catch it and the interpreter dies with no traceback, so a caller cannot even detect the gap to route around it. (Same abort string as #6563, though that one is on an NVIDIA host via nabla's autodiff; I posted a `max.graph`-only reproducer there.)

### 2. Tesla T4 / sm_75: a *mapped* cuDNN status — `CUDNN_STATUS_ALLOC_FAILED` on every shape

On CUDA the symbol load succeeds and the kernel dispatches, but fails at `conv_transpose_mogg`:

```
cuDNN call failed with status CUDNN_STATUS_ALLOC_FAILED
```

Unlike this issue's `invalid cudnnStatus_t entry`, here MAX *can* map the status — so if the A10 case is version-skew on the enum, the T4 case looks like the dispatch getting a real (but implausible) failure back. I swept five `(stride, output_padding)` pairs, so kernel size `k = 2*stride` covers 16, 10, 8, 4, 6, and the matching CPU run of the identical graph passes every time:

```
case=0 stride=8 output_padding=0 k=16  cpu -> (1, 16, 136) nan_inf=0 PASSED | gpu -> CUDNN_STATUS_ALLOC_FAILED
case=1 stride=5 output_padding=1 k=10  cpu -> (1, 16,  86) nan_inf=0 PASSED | gpu -> CUDNN_STATUS_ALLOC_FAILED
case=2 stride=4 output_padding=0 k=8   cpu -> (1, 16,  68) nan_inf=0 PASSED | gpu -> CUDNN_STATUS_ALLOC_FAILED
case=3 stride=2 output_padding=0 k=4   cpu -> (1, 16,  34) nan_inf=0 PASSED | gpu -> CUDNN_STATUS_ALLOC_FAILED
case=4 stride=3 output_padding=1 k=6   cpu -> (1, 16,  52) nan_inf=0 PASSED | gpu -> CUDNN_STATUS_ALLOC_FAILED
```

**The uniformity is the diagnostic bit.** A workspace-sizing bug tied to shape should affect some kernel sizes and not others; every size from 4 to 16 fails identically. And the tensors are tiny — 16 channels, `seq_len=16`, batch 1 — on a **fresh 15 GB T4 with nothing else resident**. An allocation failure at that size suggests the requested workspace isn't proportional to the op's real needs. The alternative I can't rule out from outside is an `sm_75`-wide incompatibility in whichever cuDNN algorithm MAX's heuristic picks for transposed conv.

**CPU is correct on both platforms** (5/5, identical shapes on M1 and T4, zero NaN/Inf), which localizes all of this to GPU dispatch rather than op semantics or the layout mapping.

Reproducer is the same shape as the one in this issue — a `max.graph`-only forward `ops.conv2d_transpose`; full script posted in #6563. Each case needs its own subprocess plus `python -u`, since the Metal abort is fatal and would otherwise hide the remaining cases.

Worth noting the op's source already has a commented-out CPU fallback that would turn all of the above into slow-but-working rather than crashing:

```python
# TODO(GEX-2043): Add support for GPU kernel for conv_transpose and remove manual transfers
# original_device = x.type.device
# x = x.to(DeviceRef.CPU())
# filter = filter.to(DeviceRef.CPU())
```

(`max/python/max/graph/ops/conv_transpose.py:121`)

### Separately found: the `bias=` path double-permutes an NHWC output

Not GPU-related and not the subject of this issue, but same op and same file, so noting it here rather than opening another issue. Passing `bias=` produces a **malformed shape** — `out_channels` lands on two axes — while `bias=None` on the identical op is correct. Device-independent, visible in symbolic shape inference with no accelerator:

```text
x_nhwc shape:        [Dim(1), Dim(1), Dim(12), Dim(8)]
filter shape:        [Dim(1), Dim(10), Dim(6), Dim(8)]
no-bias out shape:   [Dim(1), Dim(1), Dim(66), Dim(6)]   <- correct: [B, H=1, T_out, C_out]
with-bias out shape: [Dim(1), Dim(6), Dim(66), Dim(6)]   <- wrong: C_out on axes 1 and 3
```

Cause looks like the bias branch assuming an NCHW output when the output is already NHWC:

```python
if bias is not None:
    # Convert from NCHW to NHWC for bias broadcasting.
    out = permute(out, [0, 2, 3, 1])
    out = add(out, bias)
    # Convert back from NHWC to NCHW.
    return permute(out, [0, 3, 1, 2])
```

The two permutes aren't a round-trip for an NHWC output, so the bias broadcasts against the wrong axis: `[1,1,66,6]` → permute → `[1,66,6,1]` → `add` bias `(6,)` broadcasting against the trailing `1` → `[1,66,6,6]` → permute → `[1,6,66,6]`, exactly what's observed. Workaround is `bias=None` plus adding the bias manually afterwards.

This also matches the docstring/reality mismatch about the output layout that makes the no-bias path need a `squeeze(y, 1)` instead of the documented NCHW handling — a fix here probably wants the docstring corrected too. Happy to split this into its own issue if you'd prefer it tracked separately.

### Environment

```
MAX 26.5.0, Mojo 1.0.0 (ed45d567), modular 26.5.0 (conda, https://conda.modular.com/max)

Platform A -- Apple M1, macOS 26.6.2 (25G83), Xcode 26.6 (17F113),
              Metal compiler 32023.883, Python 3.14.7
Platform B -- NVIDIA Tesla T4 (sm_75, 15 GB), Google Colab default T4 runtime
              (no more specific CUDA driver/toolkit version captured for this run)
```

**Comment by neuregex** [comment]
The `ALLOC_FAILED` in the T4 run above is most likely a `BAD_PARAM`. In `max/kernels/src/_cudnn/infer.mojo` the `*_V9` status values look like the old codes × 1000, but cuDNN 9 numbers them differently: 2000 is `CUDNN_STATUS_BAD_PARAM` (the binding prints it as `CUDNN_STATUS_ALLOC_FAILED`), 3000 is `NOT_SUPPORTED`, 5000 is `EXECUTION_FAILED`, and allocation failures are 4003/4004. I checked each code against `cudnnGetErrorString` on cuDNN 9.26.

It also means the OOM retry in `max/kernels/src/nn/conv/conv.mojo` (~line 6662) fires on `BAD_PARAM` and not on a real OOM.

@BradLarson does your local patch already cover this? If not, I'm happy to send a small PR fixing the enum and the retry check.
